### DISCUSSION TITLE
Is it possible to use an aspect several times on the same type?

### DISCUSSION THREAD

**Discussion by YBAZAN** [metalama/Metalama#99] [Q&A] [answered]
Hi,

I would like to generate properties for Avalonia like https://github.com/HavenDV/DependencyPropertyGenerator does.
To be able to do that, I would need to set an same aspect several times.

Here the link to what I'm able to do by now: 
https://try.metalama.net/#N4IgZiBcDaoHZRABQE4HsDmKCGBbAdAMYDOIANCIYgDpzQDKALgJ4A2ApgCapoAO7KFgAoW-NGCHFGKAJZwMASjIACaiAByedmoUBdWgxYdu6foOYjmYiVNnylqjTMIBrTbm3llAEXZhsAK6sjABq2KwB7MoAvI4AKmiMaDr6dExsXDxmwqLs4kKcaAEARhwOagDq7DIYABaMaiq+-kGh4ZExygAcAAycerRyjAJw4cqErNjExMo8WHjKkMoAggBu4WhwMtgA8sUAVuyEDXDAtAC+tLS8JazO45PTyunGWQIsADxxAHzKwJdwa63e4TKYzNYbLbYN7mP60ZQI5Q3Ur3KTYRj3F6ZUzvZhfMIRdi-ABK7AwMikAi+OwA7nABCo4gTIt8hPDERzbHIMMpRh4yOyOQime0opw-IFgsyorFxS1ggBCBSqOBCxE02oCKJxWn0lBLCGsTbbPaHY6CuGqtXKFDsRgBFCq+k055GbF8XH40WshQAbgtAIBQJRhAeYJW6yNUNNRxOZytCORd1DcWUAHE7dKvqysSYPeZs0iceYFBb49abXaHaq5ZLGP6E8og42k-dVmgZJxnpnRdmhLmYZ8fkX8yxGcpI5EVEMizI0LIWJ0AIylxvlxGB2jkEAuRCMEDnMjwRDLYj8Y5EUgUKiQEABYjc57MSkEOLsAAe9do98fAFk7eEeDYPgABiOAeDS84uPgp7nowxANj+8jKP+jCAbgwFgVokEoNBADCaDiohD7Iah6GYeB7A4fhhHsPg9DMHAaHvgAQgEMisOKKAIVcdDLIw0gyMUATDAAqsQ2AYOwQj8YJwnDHE2AoFJ8H4HhjzECoyysEaNK-q0Mi8BwnTSJEAxwK2obEOw4RcEiSkYmMoJPAOxYsLJsjyVESxxFY7CwbGtDromwKhq5o7MB5QkidJvn8MouQqFyyF8uwq4csFQpxTKCV+Q2FbuDlqX5RuvEcgA9JVygfMQAS4BhKDMN8GbwQlmq5fF4htVEvBucwCVoMoUl6uidEfOVtX1UpTUWpZyjZfNfl-ENdq+k2ZWIpV5XVZNDVNS1MyMO1qXKF1R09X1A0rSNwz4ONu3Td8s2hcoyU8oVy0qWtzYVVVNV1XtzV2jM86vcD3XKLWrQTqKd0TQDj3PSGp0HLGAD8PgSq00qfatYP1utgKNtAb64EZo2pByvUyOswyQ4xeDOBjLXDCgpPk8MQgAJIgTI7CccoYB85x6WIplHK2vajrKB4aH4CxUx0S10pCEL-P9CVCI-YiJPsGTkzDJTiLU7TYoM7gTPdgJAjswb0mcObbaiioPPC12asi2WFoS1W0uy8BcS1BS9E9oSqtuyok7sCoPR+gGG0Ilt1VyJqC6FIQ5VPS2L1oKsAiyOKE4dl2bEcZwAXHNzFeMKXnFUlzhWcNlvzCWXAiiwi4uIsUivy+xnHV0Ird1ygcdExW6woILbsLdE3vWkI3ON9lCjZSB2DHPOzD4C12WWNY-Zunm2R4t8Cjn-gFQyEd2XLMpdXsExxBCCvDbzwik+vZ0zqunYGC11xNkjYKwABI1CGmNNCPq+BSTkkpCgD4wBh5cXwIpZSdp8CFUPH8bKmCtDnFZKlJY1ASEgGAFgkhagdCayFDIMAyghDNDrDjCkvJEi8iCKwDuQou5qjoQw2esRcj5DeufV6MFeD8DgJwIQYCvBQylKKYhpDgBMOxqKS4pDqHvyFPzay4jliSMfjIuRKgFFtEJEsVRWNFGEkuCAMeFYAQVmIBIqRMidDaOAYiHRn8PZdliMggQMFOCrGcHRLmTF0CcACIQdgvN1ZD37igtBKkVByPIfgwczBGjT3VtlAU3i1TIISZxJY9DohZwrBlHRFYwD4C5sQJg6J7hCJQJEGh1TET1OWIQOJ0whIcWvv1WIvT+kPmKEMlg+AkChU6V0wWDStiOTuAALwEAAUXfL1dgAzNidC2TsvZcB16b0ajMpS1lJCoLQEwP+Qhz7zLVOcBQ+BfCghwBiTYb8inKCCSgEJYS4lLOkIRWJ7BslJLbgC1JdpCkLN5FoJYGSKHkFqYiFSttRpLFSvkFmNs9YczSvChZ1lGBYuGDirQ+R6B2lZhS4l6KETIOyUsXgMQqkLN4dU3gqClqxDXhvJI5zd5+Rfn5RxCzsFKQwMQHFVFlr+KbJKrWFxty7lvNgA8uhzhAA

Another question: is it possible to introduce a field with the `readonly` keyword?

**Comment by svick** [comment, marked as answer]
> Is it possible to use an aspect several times on the same type?

Yes, by accessing `builder.AspectInstance.SecondaryInstances`. For example, [like this](https://try.metalama.net/#N4IgZiBcDaoHZRABQE4HsDmKCGBbAdAMYDOIANCIYgDpzQDKALgJ4A2ApgCapoAO7KFgAoW-NGCHFGKAJZwMASjIACaiAByedmoUBdWgxYdu6foOYjmYiVNnylqjTMIBrTbm3llAEXZhsAK6sjABq2KwB7MoAvI4AKmiMaDr6dExsXDxmwqLs4kKcaAEARhwOagDq7DIYABaMaiq+-kGh4ZExygAcAAycerRyjAJw4cqErNjExMo8WHjKkMoAggBu4WhwMtgA8sUAVuyEDXDAtAC+tLS8JazO45PTyunGWQIsADxxAHzKwJdwa63e4TKYzNYbLbYN7mP60ZQI5Q3Ur3KTYRj3F6ZUzvZhfMIRdi-ABK7AwMikAi+OwA7nABCo4gTIt8hPDERzbHIMMpRh4yOyOQime0opw-IFgsyorFxS1ggBCBSqOBCxE02oCKJxWn0lBLCGsTbbPaHY6CuGqtXKFDsRgBFCq+k055GbF8XH40WshQAbgtAIBQJRhAeYJW6yNUNNRxOZytCORd1DcWUAHE7dKvqysSYPeZs0iceYFBb49abXaHaq5ZLGP6E8og42k-dVmgZJxnpnRdmhLmYZ8fkX8yxGcpI5EVEMizI0LIWJ0AIylxvlxGB2jkEAuRCMEDnMjwRDLYj8Y5EUgUKiQEABYjc57MSkEOLsAAe9do98fAFk7eEeDYPgABiOAeDS84uPgp7nowxANj+8jKP+jCAbgwFgVokEoNBADCaDiohD7Iah6GYeB7A4fhhHsPg9DMHAaHvgAQgEMisOKKAIVcdDLIw0gyMUATDAAqsQ2AYOwQj8YJwnDHE2AoFJ8H4HhjzECoyysEaNK-q0Mi8BwnTSJEAxwK2obEOw4RcEiSkYmMoJPAOxYsLJsjyVESxxFY7CwbGtDromwKhq5o7MB5QkidJvn8MouQqFyyF8uwq4csFQpxTKCV+Q2FbuDlqX5RuvEcgA9JVygfMQAS4BhKDMN8GbwQlmq5fF4htVEvBucwCVoMoUl6uidEfOVtX1UpTUWpZyjZfNfl-ENdq+k2ZWIpV5XVZNDVNS1MyMO1qXKF1R09X1A0rSNwz4ONu3Td8s2hcoyU8oVy0qWtzYVVVNV1XtzV2jM86vcD3XKLWrQTqKd0TQDj3PSGp0HLGAD8PgSq00qfatYP1utgKNtAb64EZo2pByvUyOswyQ4xeDOBjLXDCgpPk8MQgAJIgTI7CccoYB85x6WIplHK2vajrKB4aH4CxUx0S10pCEL-P9CVCI-YiJPsGTkzDJTiLU7TYoM7gTPdgJAjswb0mcObbaiioPPC12asi2WFoS1W0uy8BcS1BS9E9oSqtuyok7sCoPR+gGG0Ilt1VyJqC6FIQ5VPS2L1oKsAiyOKE4dl2bEcZwAXHNzFeMKXnFUlzhWcNlvzCWXAiiwi4uIsUivy+xnHV0Ird1ygcdExWR3B7X5dnrGXNwGicCENJw9cWP3uImA842YQtTKEI6woMociL8vx+qqvAgwbPxzz6fdH0EcmycNNd9oUv7DEB3Qpd2qQj9m6PM2RIoCU8jFBQJ937L2vnBBQfcy7VzftgD+Q9+5r01kKAEHJtaJlkKbIunZlDT0QQvKB0kubV2nvXRuzdlCX1Hl7RsHJD6CzdgtaIG8-7cxoX5BQ2UQLYGOPOZg+AWrZUsNYABGQgGem+AoeR+AKgyCOtlZYyk6rsCYsQIQ2V15MMRCwmYsRnSujsBgKhKA2T6LVAAEjUIaY00I+r4FJOSSkKAPjAHofgRSyk7T4EKoeP42UAlaHOKyVKSxqDRJAMAQJ0S1A6AwRyGQYB97NDrDjCkvJEi8iCKwb+GVOFClSfvdhsRcj5DevI16MFeD8DgJwIQdivBQylKKKJMTgAZOxqKS4MSknFI5PzaytTlj1M0U0lpKg2ltEJEsbpWN2mEkuCAMeFYsHWmIHUhpTSdCDOsSqCsLCPZdliN45YnBVjODovPaQhEAjL15urVBbcUA+KUipFQLS4lhMHMwRorD1bZQFIcjkq9nmcSWGk6IWcKxFLBdaMA+AubECYOie4FSUCRGSfChEyLliEGXtMISHFlH9ViIS4lD5ihkpYPgJAoVcV4uRfPZR2w7gAC8BAAFF3y9U-g+TYnQ+UCpJZsARQjGoMqUtZSQPi0BMDMUIeRzLMFwN8KCHAGJNgNmKRcq5NyUVMXQJwR57B-mvJHh8vxjBQV4oRJE5QPz4nkCGYiFSttRpLFSvkFmNs9YczSvah11lGBeuGD6rQ+RH7WzZoGu2Sh3UIlXv8pYvAYhwodb-eFvAfFLViPwwRSRpViL8jo3harsEqE+cQH1VFlqnKbOs0qcBzjbl3LebAB5dDnCAA):

```c#
public override void BuildAspect(IAspectBuilder<INamedType> builder)
{
    base.BuildAspect(builder);

    this.BuildAspectInstance(builder);

    foreach (var instance in builder.AspectInstance.SecondaryInstances)
    {
        ((StyledPropertyAttribute)instance.Aspect).BuildAspectInstance(builder);
    }
}

private void BuildAspectInstance(IAspectBuilder<INamedType> builder)
{
    // your original BuildAspect code here
}
```

> is it possible to introduce a field with the `readonly` keyword?

Yes, if you use a template field, e.g.:

```c#
[Template]
private readonly object Field;

...

var field = builder.Advice.IntroduceField(builder.Target, nameof(Field),
    buildField: f =>
    {
        f.Name = $"{Name}Property";
        f.Type = fieldType;
        f.IsStatic = true;
        f.Accessibility = Accessibility.Public;
        f.InitializerExpression = ExpressionFactory.Parse(s.ToString());
    }).Declaration;
```

Though the inability to directly set `Writeability` on an `IFieldBuild` feels limiting to me. I'll see if we can improve that.
